### PR TITLE
fix(polys): don't unify characteristics

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Partial fix for #25653 by disallowing any unification of domains with different characteristic.

#### Brief description of what is fixed or changed

Previously:
```python
>>> GF(3).unify(ZZ)
ZZ
```
With the PR:
```python
>>> GF(3).unify(ZZ)
...
UnificationFailed: Cannot unify GF(3) with ZZ 
```

Also adds more consistent handling of `.has_CharacteristicZero` and `.characteristic()` for different domains.

#### Other comments

There was one place in the code

#### Release Notes

* polys
    * BREAKING CHANGE: Unification of domains like `GF(p)` with `ZZ` or other domains would previously result in `ZZ` (or any other domain). This now results in a UnificationFailed error. Previously unification like e.g. `GF(2).unify(GF(3))` would unify to the large characteristic but this now results in UnificationFailed.

### CODE REVIEW

File: sympy/polys/numberfields/primes.py [hunk 1]
```diff
@@ -396,7 +396,7 @@ def _compute_test_factor(p, gens, ZK):
     # predicts that such an element must exist, so nullspace should
     # be non-trivial.
     x = B.nullspace()[0, :].transpose()
-    beta = ZK.parent(ZK.matrix * x, denom=ZK.denom)
+    beta = ZK.parent(ZK.matrix * x.convert_to(ZZ), denom=ZK.denom)
```
Comment (oscarbenjamin) [inline sympy/polys/numberfields/primes.py:399]: @skieffer this change preserves existing behaviour. Let me know if something else should be done here.

File: sympy/simplify/tests/test_ratsimp.py [hunk 2]
```diff
@@ -74,5 +74,5 @@ def test_ratsimpmodprime():
         y/2
 
     a = (x**5 + 2*x**4 + 2*x**3 + 2*x**2 + x + 2/x + x**(-2))
-    assert ratsimpmodprime(a, [x + 1], domain=GF(2)) == 1
-    assert ratsimpmodprime(a, [x + 1], domain=GF(3)) == -1
+    # assert ratsimpmodprime(a, [x + 1], domain=GF(2)) == 1
+    # assert ratsimpmodprime(a, [x + 1], domain=GF(3)) == -1
```
Comment (oscarbenjamin) [inline sympy/simplify/tests/test_ratsimp.py:78]: These examples highlight a problem with disallowing any unification between `GF(p)` and `ZZ`:
```python
In [1]: p = Poly(1, x, modulus=2)

In [2]: p
Out[2]: Poly(1, x, modulus=2)

In [3]: p * 2
...
UnificationFailed: Cannot unify GF(2) with ZZ
```
There are probably ways to work around this but at least it shows why being able to unify `GF(p)` and `ZZ` makes sense in some situations. Here the problem is that the `2` is coerced into `Poly(2, x, domain=ZZ)` before the product is computed. Then the product tries to unify the two `Poly` and fails if `GF(2)` cannot unify with `ZZ`.

Comment (oscarbenjamin) [reply sympy/simplify/tests/test_ratsimp.py:77]: In this example I don't think it matters which way the unification happens as long as they do unify to something.

Comment (oscarbenjamin) [reply sympy/simplify/tests/test_ratsimp.py]: I fixed this by special casing `Poly * int/Integer`.

**Comment by oscarbenjamin** [comment]
This is ready for review.

File: sympy/polys/domains/tests/test_domains.py [hunk 3]
```diff
@@ -570,6 +580,17 @@ def test_Domain_get_exact():
     assert QQ.frac_field(x, y).get_exact() == QQ.frac_field(x, y)
 
 
+def test_Domain_characteristic():
+    for F, c in [(FF(3), 3), (FF(5), 5), (FF(7), 7)]:
+        for R in F, F[x], F.frac_field(x), F.old_poly_ring(x), F.old_frac_field(x):
```
Comment (smichr) [inline sympy/polys/domains/tests/test_domains.py:585]: R is not used in this loop interior -- is something missing?

File: sympy/polys/polytools.py [hunk 4]
```diff
@@ -71,6 +71,9 @@ def wrapper(f, g):
         g = _sympify(g)
         if isinstance(g, Poly):
             return func(f, g)
+        elif isinstance(g, Integer):
```
Comment (smichr) [inline sympy/polys/polytools.py:74]: I assume this is where you special cased Integer (as you referred to in how you fixed the previously commented out tests). Can `g` be an int or is that always sympified before it gets here?

Comment (oscarbenjamin) [reply sympy/polys/polytools.py:74]: `g` is sympified 3 lines up from here.

File: sympy/polys/domains/tests/test_domains.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/polys/domains/tests/test_domains.py:585]: Ah, good point. Fixed.

**Comment by oscarbenjamin** [comment]
Thanks!

File: sympy/polys/numberfields/primes.py (hunk 1 above)
Comment (skieffer) [reply sympy/polys/numberfields/primes.py:399]: Thanks for the question, and I'm sorry I wasn't able to respond in time. But your fix looks good. Thanks for this.

Comment (oscarbenjamin) [reply sympy/polys/numberfields/primes.py:399]: As the author of the only code in the whole of SymPy that depended on the previous behaviour do you have any opinions about `GF(3).unify(ZZ)` and the general change in this PR that disallows it?

Comment (skieffer) [reply sympy/polys/numberfields/primes.py:399]: Good question. I'll comment on the main thread, #25653.

**Pull request merged**